[XenosF] iP

xenosf opened the pull request:
# Dude

```
   _|    _| _    O
  (_||_|(_|(/_  /İ\
  ------------  ```
```

>  *“Know yourself, and you will win all battles.”* – Sun Tzu

Dude frees your mind of having to remember things you need to do ~~such as your CS2103 weekly tasks~~, so you can better know yourself. It's:

* `Text-based`
* Easy to learn
* 🔥🔥 ***BLAZING FAST*** 🔥🔥

Best of all, it's **FREE**!

## Features

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
    public static void main(String[] args) {
        Dude dude = new Dude("./data/dude.txt");
        dude.run();
    }
```

## How to use

1. Download the latest `.jar` file [here](https://github.com/xenosf/ip/releases).
2. Copy the jar file into an empty folder.
3. Open a command window in that folder.
4. Run the command `java -jar Dude.jar` (i.e., run the command in the same folder as the jar file).

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

xenosf committed:
rename duke to dude; set up greeting and exit messages

xenosf committed:
add .class files to gitignore

xenosf committed:
add echo user command feature & bye command

xenosf committed:
separate command and text parsing

xenosf committed:
move messages out of main function

xenosf committed:
move input parsing out of main function

xenosf committed:
add task list add and display feature

xenosf committed:
refactor tasks add and list to separate methods; add 0 tasks msg

xenosf committed:
add mark and unmark task feature

xenosf committed:
streamline console print function calls

xenosf committed:
change getTaskString to override toString instead

xenosf committed:
remove isdone getter

xenosf committed:
add different types of tasks: deadline, event, todo

xenosf committed:
change task type icon formatting

xenosf committed:
change task date display formatting

xenosf committed:
add todo task, deadline task, event task commands

xenosf committed:
change initial ascii art

xenosf committed:
fix bug in event end check

xenosf committed:
edit error message to fit in lines

xenosf committed:
implement automated testing

xenosf committed:
add javadoc for Dude class

xenosf committed:
add DudeException parent class

xenosf committed:
implement invalid command exception

xenosf committed:
convert exception messages to actual exceptions

xenosf committed:
add param for index

xenosf committed:
refactor tasks to use ArrayList

xenosf committed:
add delete task method; refactor get task to use arraylist

xenosf committed:
implement delete feature

xenosf committed:
split and refactor mark and unmark

xenosf committed:
convert command to lowercase before checking

xenosf committed:
use enums for commands

xenosf committed:
convert all dude errors to use exceptions

xenosf committed:
Add save data to gitignore

xenosf committed:
Implement save feature

xenosf committed:
Merge pull request #1 from xenosf/branch-Level-7

Merge Branch level 7

derekjxtan commented on src/main/java/Dude.java:
````diff
@@ -0,0 +1,419 @@
+      "(_||_|(_|(/_  /İ\\\n" +
+      "------------  ```\n";
+  static String border = "-----------------------------------------------------------\n";
+  static String hello = logo +
````
Maybe these strings could be made final and prefixed with MESSAGE.

derekjxtan commented on src/main/java/Dude.java:
```diff
@@ -0,0 +1,419 @@
+   * @throws DudeException if parsing task from command fails
+   */
+  public static void parseDelete(String input) throws DudeException {
+    int taskToDelete = parseTaskIndexCommand(input);
```
maybe this can be renamed to taskIndexToDelete so that its clearer that it is a index not a task?

derekjxtan commented on src/main/java/Dude.java:
```diff
@@ -0,0 +1,419 @@
+   * @param input command.
+   * @throws TaskDescriptionMissingException if task description is missing
+   */
+  public static void parseTodo(String input) throws TaskDescriptionMissingException, SaveFileException {
```
maybe its better to name this in such a way that others know this function does the creation on top of parsing? Or maybe splitting up the logic into 2 functions might be better?

derekjxtan commented on src/main/java/DudeCommand.java:
```diff
@@ -0,0 +1,14 @@
+ * Commands to be entered into Dude.
+ */
+public enum DudeCommand {
+  list,
```
I think enums should be capitalized since they are like constants. What do you think?

derekjxtan commented on src/main/java/ToDoTask.java:
```diff
@@ -0,0 +1,73 @@
+   * @return Task instance
+   * @throws InvalidTaskDataException if data is not in the expected format
+   */
+  public static ToDoTask fromData(String data) throws InvalidTaskDataException {
```
this could use the fromData method in Task so reduce the amount of duplicated code

derekjxtan reviewed (commented):
Overall code standard is good! Keep up the good work :)

xenosf committed:
Update test scripts to delete saved data

xenosf committed:
Add proper date time parsing

xenosf committed:
Merge branch 'branch-Level-8'

Implement date and time input and processing for tasks that store
dates/times.

xenosf committed:
Refactor storage out into its own class

xenosf committed:
Refactor UI out into its own class

xenosf committed:
Set file path as final

xenosf committed:
Refactor commands into individual classes

xenosf committed:
Fix index out of bounds error when saving tasks

xenosf committed:
Refactor parser functions out into separate class

xenosf committed:
Refactor task list out into separate class

xenosf committed:
Refactor main Dude functionality

xenosf committed:
Merge branch 'branch-A-MoreOOP'

Refactor Dude into several smaller classes.

xenosf committed:
Organize classes into packages

xenosf committed:
Merge branch 'branch-A-Packages'

Organize classes into packages.

xenosf committed:
Merge branch 'add-gradle-support'

xenosf committed:
Add checkstyle plugin and corresponding config

xenosf committed:
Add unit tests for datetime parsing

xenosf committed:
Add unit tests for task list get & remove

xenosf committed:
Rename JAR export base name to dude

xenosf committed:
Add task search feature

xenosf committed:
Display correct task index in search results

xenosf committed:
Fix test scripts to use package

xenosf committed:
Update test cases to test find function

xenosf committed:
Add and reformat header comments

xenosf committed:
Reformat code to follow coding standard.

xenosf committed:
Merge branch 'branch-Level-9'

Implement task search function.

xenosf committed:
Merge branch 'branch-A-JavaDoc'

Add and improve JavaDoc documentation.

xenosf committed:
Merge branch 'branch-A-CodingStandard'

Reformat code to follow coding standard.

xenosf committed:
Further reformat code to follow coding standard.

xenosf committed:
Make message constants private

xenosf committed:
Update README.md

rocketninja7 commented on src/main/java/DeadlineTask.java:
```diff
@@ -0,0 +1,84 @@
+    String[] splitData = data.split(DELIMITER_REGEX, 4);
+
+    if (splitData.length < 4) {
+      throw new InvalidTaskDataException();
```
Maybe the indentation could be 4 spaces to comply with the coding standards? I noticed the same in other places.

rocketninja7 commented on src/main/java/DeadlineTask.java:
```diff
@@ -0,0 +1,84 @@
+   * @throws InvalidTaskDataException if data is not in the expected format
+   */
+  public static DeadlineTask fromData(String data) throws InvalidTaskDataException {
+     /*
```
I like how you have descriptive documentation to show the expected format. However, I was wondering if this should be part of the JavaDocs?

rocketninja7 reviewed (commented):
Looks good.
Some minor changes required.

xenosf committed:
Add JavaFX dependencies to Gradle

xenosf committed:
Implement barebones GUI

xenosf committed:
Implement interaction and improve GUI

xenosf committed:
Remove testing for text-based UI

xenosf committed:
Adapt Dude to use GUI

xenosf committed:
Reformat code

xenosf committed:
Use FXML for GUI instead

The GUI is defined using Java code.

Using Java code to define the GUI is tedious and makes the code more interdependent.

Using FXML keeps the code more concise, encapsulated, and easier to maintain.

Let's convert the GUI code to use FXML. Doing so will better separate UI and logic considerations.

xenosf committed:
Merge branch 'branch-Level-10'

xenosf committed:
Enable running with assertions in Gradle

Assertions are not enabled in Gradle.

It is recommended to use assertions liberally in our code to catch bugs.
By default, assertions are disabled when running with Gradle, which causes false
assumptions that the assertions work.

Let's enable assertions when running the program using Gradle.

xenosf committed:
Document assumptions using assertions

The code makes important assumptions that may not hold true due to bugs.

If these assumptions do not hold true, the program may not work as expected.

Let's use assertions to ensure that
these important assumptions hold true.

Using Java assertions allows for a lightweight way to ensure important assumptions in the code hold.

xenosf committed:
Improve code quality by adding comments and reorganising code

xenosf committed:
Reformat code

xenosf committed:
Remove javadoc throws for method that does not throw error

xenosf committed:
Add assertions to check that properties are not null

xenosf committed:
Disable processing of blank user input

xenosf committed:
Merge pull request #3 from xenosf/branch-A-Assertions

Add Java assertions to document important assumptions in code

xenosf committed:
Merge branch 'master' into branch-A-CodeQuality

xenosf committed:
Rephrase javadoc overview

xenosf committed:
Remove unneeded .toString() call

xenosf committed:
Add @@author tag for reused launcher code

xenosf committed:
Merge pull request #4 from xenosf/branch-A-CodeQuality

Improve code quality

xenosf committed:
Disable send button when input is blank

xenosf committed:
Modify GUI spacing to look better

xenosf committed:
Modify GUI to distinguish bot and user messages by color

xenosf committed:
Remove unused bye command

xenosf committed:
Add reuse tag for reused solution

xenosf committed:
Implement task sorting functionality

xenosf committed:
Clean up code

xenosf committed:
Clean up code

xenosf committed:
Merge branch 'branch-C-Sort'

xenosf committed:
Add product screenshot

xenosf committed:
Crop product screenshot

xenosf committed:
Fix file separator issue on Windows

xenosf committed:
Write user guide

xenosf committed:
Edit README.md for clarity

xenosf committed:
Create _config.yml for user guide theme

xenosf committed:
Change docs theme

xenosf committed:
Add window title

xenosf committed:
Fix documentation sort syntax

xenosf committed:
Improve GUI

xenosf committed:
Update documentation to reflect new GUI

xenosf committed:
Update documentation visuals

xenosf committed:
Remove theme

xenosf committed:
Update documentation for clarity

xenosf committed:
Clean up code to match coding standard
